A "supported-releases" page to give a support timeline (#452)
cert-manager/website: content/en/docs/installation/supported-releases.md

<!--
Inspired by https://istio.io/latest/about/supported-releases/
-->

This page lists the status, timeline and policy for currently supported
releases.

## Support status of cert-manager releases

In the below table, each version is clickable and will direct you to the
corresponding GitHub milestone. The `~` sign is used when the date is
uncertain and might change; the actual release date might be pushed a few
weeks later. The "EOL" abbreviation above stands for End Of Life.

| Version   | Support |     Release Date     |         EOL          | Supported Kubernetes versions |
| --------- | ------- | :------------------: | :------------------: | :---------------------------: |
| master    | No      |                      |                      |                               |
| [v1.3][]  | N/A     | _[~Apr 8, 2021][tl]_ |  _[~Aug 2021][tl]_   |                               |
| [v1.2][]  | **Yes** |     Feb 10, 2021     |  _[~Jun 2021][tl]_   |          1.16 → 1.20          |
| [v1.1][]  | **Yes** |     Nov 24, 2021     | _[~Apr 8, 2021][tl]_ |          1.11 → 1.20          |
| [v1.0][]  | No      |     Sep 2, 2020      |     Feb 10, 2021     |          1.11 → 1.20          |

> | [v0.16][] | No      |     Jul 23, 2020     |     Nov 24, 2021     |          1.11 → 1.20          |

Reviewer: Either the EOL for 0.16 is incorrect or it should be marked as "Supported"

[v1.3]: https://github.com/jetstack/cert-manager/milestone/24
[v1.2]: https://cert-manager.io/docs/release-notes/release-notes-1.2
[v1.1]: https://cert-manager.io/docs/release-notes/release-notes-1.1
[v1.0]: https://cert-manager.io/docs/release-notes/release-notes-1.0
[v0.16]: https://cert-manager.io/docs/release-notes/release-notes-0.16
[v0.15]: https://cert-manager.io/docs/release-notes/release-notes-0.15
[v0.14]: https://cert-manager.io/docs/release-notes/release-notes-0.14
[v0.13]: https://cert-manager.io/docs/release-notes/release-notes-0.13
[v0.12]: https://cert-manager.io/docs/release-notes/release-notes-0.12
[v0.11]: https://cert-manager.io/docs/release-notes/release-notes-0.11

You can find available releases on the [releases
page](https://github.com/cert-manager/cert-manager/releases). You can find
the release notes for each minor release
[here](https://cert-manager.io/docs/release-notes/), and the upgrade
instructions are
[here](https://cert-manager.io/docs/installation/upgrading/). The
cert-manager release process is documented on the [release-process
page](https://cert-manager.io/docs/contributing/release-process/).

## Support policy

> In the below section, we refer

Reviewer 2: Unfinished.

Author: Fixed! 😇

   using the
   [cert-manager-dev](https://groups.google.com/g/cert-manager-dev) Google
   group.

Around once every two months, we build a minor release, e.g., `v1.4`,
denoted `N` in the below table. Around the same time, we issue one backport
patch release, e.g. `v1.3.1`, that we denote `N-1` in the below table. This
backport patch release contains the bugs that we identified as "important
bugs", such as bugs preventing the upgrade to an earlier version.

Regular bugs will only be fixed on the next release of cert-manager (`N+1`
in the below table).

| Severity      | Examples             |  Is previous release patched?  |  Is current release patched?   | Or does it wait until next release? |
| ------------- | -------------------- | :----------------------------: | :----------------------------: | :---------------------------------: |
| Critical bug  | [#2857][], [#3393][] |       Yes, patched ASAP        |       Yes, patched ASAP        |                 No                  |
| Important bug | [#3644][]            | Yes, released every two months | Yes, released every two months |                 No                  |

> | Regular bug   |                      |               No               |               No               |                 Yes                 |

Reviewer 3: I don't think there is a clear distinction here between these 3, nor what constitutes what.

Can we simplify this a bit?

1. Security issue
2. Critical bug (functionality that *did* work in the release before but no longer works/something has broken)
3. Long-standing bugs (bugs that have been acknowledged and have been known for a long time, and there are well known workarounds)

I think (1) should trigger a patch release for _all_ supported releases as soon as the patch is ready

(2) should be back-ported, but not immediately trigger a release (i.e. a patch release will be made for all supported releases according to the regular release schedule)

(3) I think is probably the hardest to define case - I think longstanding bugs typically have been around for a while because the fix is _non trivial_ (i.e. requires some heavy refactoring work, or some redesign). These sorts of patches are usually _not_ so acceptable for backporting, as they introduce large numbers of LoC changed. In these instances, I think we should not backport the patches *at all* as it could risk the stability of user clusters due to their size/complexity

Author: Thank you for your comment, this is super useful!! Can I copy some of your comments into the page?  😇

| --------- | ------- | :------------------: | :------------------: | :---------------------------: |
| master    | No      |                      |                      |                               |
| [v1.3][]  | N/A     | _[~Apr 8, 2021][tl]_ |  _[~Aug 2021][tl]_   |                               |
| [v1.2][]  | **Yes** |     Feb 10, 2021     |  _[~Jun 2021][tl]_   |          1.16 → 1.20          |
| [v1.1][]  | **Yes** |     Nov 24, 2021     | _[~Apr 8, 2021][tl]_ |          1.11 → 1.20          |
| [v1.0][]  | No      |     Sep 2, 2020      |     Feb 10, 2021     |          1.11 → 1.20          |
| [v0.16][] | No      |     Jul 23, 2020     |     Nov 24, 2021     |          1.11 → 1.20          |
| [v0.15][] | No      |     May 6, 2020      |     Sep 2, 2020      |          1.11 → 1.20          |
| [v0.14][] | No      |     Mar 11, 2020     |     Jul, 23 2020     |          1.11 → 1.20          |
| [v0.13][] | No      |     Jan 21, 2020     |     May, 6 2020      |          1.11 → 1.20          |
| [v0.12][] | No      |     Nov 27, 2019     |     Mar 11, 2020     |          1.11 → 1.20          |

> | [v0.11][] | No      |     Oct 10, 2019     |     Jan 21, 2020     |          1.9 → 1.20           |

Author: I'm not sure about the supported Kubernetes versions

Reviewer 4: Is it the case that we need to sync our supported Kubernetes versions with upstream?
For example, right now when I deploy `cert-manager` on k8s v1.19 I get this in logs 
```
W0329 20:23:26.664405       1 warnings.go:67] networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
```

Maartje says [here](https://github.com/jetstack/cert-manager/pull/3499) that we can start using `networking.k8s.io/v1` once `cert-manager`'s minimum supported k8s version is v1.19 and I guess we have to do it before we start supporting v1.22, so I guess we wouldn't, for example, be able to support v1.18 and v1.22 simultaneously?
Perhaps always the latest 3 or 4 minor releases?

Reviewer: For older versions which aren't supported today, is it worth leaving out the supported k8s versions info and just putting "N/A"? There's no realistic guarantee that, say, v0.11 is gonna work on k8s 1.20, and nobody is going to be running tests to check that.

By the time this PR is merged, the supported versions will be v1.3 and v1.2 - I think we should document which versions of k8s we support for those two versions, and leave the rest blank.

Reviewer 2: My understanding is 

K8S 1.19, the support window is 1 year, so 1.19, 1.20, 1.21 will received patch updates for 1 year after their release.

K8S <= 1.18, the support window is 9 months, so 1.18, 1.17 and 1.16 are no longer supported.
(although they do still seem to be patching 1.18)

So I think we should only run our tests on 1.19, 1.20 and 1.21.

https://kubernetes.io/blog/2020/08/31/kubernetes-1-19-feature-one-year-support/

Although when I look at the tag dates for those older versions, it doesn't really match the 9 month policy:

1.16:
 * https://github.com/kubernetes/kubernetes/tree/v1.16.0 (13 Sep 2019)
 * https://github.com/kubernetes/kubernetes/tree/v1.16.15 (2 Sep 2020)

1.17:
 * https://github.com/kubernetes/kubernetes/tree/v1.17.0 (6 Dec 2019)
 * https://github.com/kubernetes/kubernetes/tree/v1.17.17 (13 Jan 2021)

1.18:
 * https://github.com/kubernetes/kubernetes/tree/v1.18.0 (25 Mar 2020)
 * https://github.com/kubernetes/kubernetes/tree/v1.18.17 (18 Mar 2021)

Reviewer 2: AKS release and supported version schedule: https://docs.microsoft.com/en-us/azure/aks/supported-kubernetes-versions#aks-kubernetes-release-calendar

GKE supported versions: https://cloud.google.com/kubernetes-engine/docs/release-notes

Reviewer 2: Kubernetes Yearly Support Period  KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-release/1498-kubernetes-yearly-support-period

Author: @wallrj Could we have a quick call about that? I feel a bit lost with all these versions and I don't know what I should be writing anymore 😅

Author: Note: our current coverage still does not test Kubernetes 1.21.

| Release branch | Prow configuration            | testgrid                | Kubernetes versions tested   | Periodicity |
| :------------: | :---------------------------- | :---------------------- | :--------------------------- | :---------: |
|     master     | [`next-periodics.yaml`][]     | [`testgrid-next`][]     | 1.16, 1.17, 1.18, 1.19, 1.20 | On each PR  |
|  release-1.3   | [`previous-periodics.yaml`][] | [`testgrid-previous`][] | 1.16, 1.17, 1.18, 1.19, 1.20 | Every night |
|  release-1.2   | N/A                           |                         | N/A                          |     N/A     |

Author: I'll resolve this comment for now, although I wish I didn't because this thread is very informative

[group]: https://groups.google.com/g/cert-manager-dev

### Security and bug fixes {#bug-fixes-support}

We back-port important bug fixes — including security fixes — to all
currently supported releases.

- [Security issues](#security-issues),
- [Critical bugs](#critical-bugs),
- [Long-standing bugs](#long-standing-bugs).

#### Security issues {#security-issues}

**Security issues** are fixed as soon as possible. They get back-ported to
the last two releases, and a new patch release is immediately created for
them.

#### Critical bugs {#critical-bugs}

**Critical bugs** include both regression bugs as well as upgrade bugs.

Regressions are functionalities that worked in a previous release but no
longer work. [#3393][] and [#2857][] are two examples of regressions.

Upgrade bugs are issues (often Helm-related) preventing users from
upgrading to currently supported releases from earlier releases of
cert-manager. [#3882][] and [#3644][] are examples of upgrade bugs.

Note that intentional breaking changes do not belong to that category.
Intentional breaking changes are changes that have been announced at least
four months prior. For example, an intentional breaking change that is

> shipped in release 1.4 will first be announced in 1.2.

Reviewer: question:

1. is four months something we've discussed and agreed on? It seems an exceptionally long time for some changes.

2. is it worth giving intentional breaking changes their own section and referring to that section in the "critical bugs" section?

Author: > Give breaking changes their own section

Good point, I'll do that.

> is four months something we've discussed and agreed on? It seems an exceptionally long time for some changes.

I have to admit that we never agreed on anything regarding this "4 months notice period". I might have picked this 😅

But yeah, I do agree that it seems to be an extremely long notice period... but that's something I would really expect from cert-manager: don't break things too fast! 😇

Thoughts? @jakexks @irbekrm @JoshVanL @wallrj @munnerz

Author: Done in b6a580f 👍

[1.3]: https://cert-manager.io/docs/release-notes/release-notes-1.3
[1.2]: https://cert-manager.io/docs/release-notes/release-notes-1.2
[1.1]: https://cert-manager.io/docs/release-notes/release-notes-1.1
[1.0]: https://cert-manager.io/docs/release-notes/release-notes-1.0
[0.16]: https://cert-manager.io/docs/release-notes/release-notes-0.16
[0.15]: https://cert-manager.io/docs/release-notes/release-notes-0.15
[0.14]: https://cert-manager.io/docs/release-notes/release-notes-0.14
[0.13]: https://cert-manager.io/docs/release-notes/release-notes-0.13
[0.12]: https://cert-manager.io/docs/release-notes/release-notes-0.12
[0.11]: https://cert-manager.io/docs/release-notes/release-notes-0.11

You can find available releases on the [releases
page](https://github.com/cert-manager/cert-manager/releases). You can find
the release notes for each minor release
[here](https://cert-manager.io/docs/release-notes/), and the upgrade
instructions are
[here](https://cert-manager.io/docs/installation/upgrading/). The
cert-manager release process is documented on the [release-process
page](https://cert-manager.io/docs/contributing/release-process/).

> ## Support policy

Reviewer: nit: we have a mixture of title case and non-title-case titles in this document. I prefer title case everywhere, but the most important thing is consistency IMO; it seems small and nitpicky but I really think consistency in title casing massively improves how the document reads. I'll raise suggestions on each title, although I don't think this is blocking.
```suggestion
## Support Policy
```

Reviewer: to be clear since this was raised in standup:

- Whether or not to use title casing is a nitpick and a stylistic choice. I think that was clear in my original comment.
- It's your PR and certainly shouldn't be blocked based on a stylistic choice.
- I _personally_ prefer title casing everywhere and I would use title casing if I were writing a document
- The most important thing by far is consistency, and this document uses a mixture of approaches to title casing. Whether we use it everywhere or nowhere in the document, either is fine. It's the mixing which is an issue.

Author: Thanks for the clarification!! I'll make sure to have the casing consistent. You are right, that's the most important bit, the rest is less important (we love the occasional bikeshedding, don't we 😇)

> ### What we mean by support

Reviewer: nit: title case 
```suggestion
### What We Mean by Support
```

Author: Fixed in [e05d5d2](https://github.com/cert-manager/website/commit/e05d5d245379f4ebc7ac4fd7fe25a4265437c4c7) (consisntent casing)

   v1.0.0                                                          ^
 Sep 2, 2020                                                       | UNSUPPORTED
------+---------------------------------------------> release-1.0  | RELEASES
       \                                                           v
        \
         \       v1.1.0
          \    Nov 24, 2021                                        ^
           ---------+-------------------------------> release-1.1  |
                     \                                             | SUPPORTED
                      \                                            | RELEASES
                       \         v1.2.0                            | = the two
                        \      Feb 10, 2021                        |   last
                         ------------+--------------> release-1.2  |   releases
                                      \                            v
                                       \
                                        \
                                         \
                                          -----------> master branch
                                                       April 1, 2021
```

> ### Technical support {#technical-support}

Reviewer: nit: title case

```suggestion
### Technical Support {#technical-support}
```

Author: Fixed in e05d5d2 (consisntent casing)

Technical assistance is offered on a best-effort basis for supported
releases only. You can request support from the community on [Kubernetes
Slack](https://slack.k8s.io/) (in the `#cert-manager` channel), using
[GitHub Discussions][discussions] or using the [cert-manager-dev][group]
Google group.

[discussions]: https://github.com/jetstack/cert-manager/discussions
[group]: https://groups.google.com/g/cert-manager-dev

> ### Security and bug fixes {#bug-fixes-support}

Reviewer: nit: title case

```suggestion
### Security and Bug Fixes {#bug-fixes-support}
```

Author: Fixed in e05d5d2 (consisntent casing)

We back-port important bug fixes — including security fixes — to all
currently supported releases.

- [Security issues](#security-issues),
- [Critical bugs](#critical-bugs),
- [Long-standing bugs](#long-standing-bugs).

> #### Security issues {#security-issues}

Reviewer: nit: title case

```suggestion
#### Security Issues {#security-issues}
```

Author: Fixed in e05d5d2 (consisntent casing)

**Security issues** are fixed as soon as possible. They get back-ported to
the last two releases, and a new patch release is immediately created for
them.

> #### Critical bugs {#critical-bugs}

Reviewer: nit: title case
```suggestion
#### Critical Bugs {#critical-bugs}
```

Author: Fixed in e05d5d2 (consisntent casing)

In the example below, the release branches `release-1.2` and `release-1.3`
are the two supported releases at the time of the release of 1.3.0, which
means that critical bug fixes 1 and 2 will only be back-ported to
`release-1.2` and not to `release-1.1`:

```diagram
   v1.2.1                                v1.2.2
------+-------------------------------------+-----------> release-1.2
       \      backport ^  backport ^        ^
        \     commit  /   commit  /         | v1.2.2 is created
         \           /           /          | along with v1.3.0
          \         /           /
           \       /           /
            \     /           /          v1.3.0
             --------master-----------------+-----------> release-1.3
                 ^           ^               \
                 |           |                \
                 |           |                 \
           critical      critical               \
           bug fix 1     bug fix 2               \
           merged to     merged to                \
           master        master                    ------> master
```

<!-- Diagram source: https://textik.com/#7c4096204b3c0ad3 -->

> #### Long-standing bugs {#long-standing-bugs}

Reviewer: nit: title case
```suggestion
#### Long-Standing Bugs {#long-standing-bugs}
```

Author: Fixed in e05d5d2 (consisntent casing)

**Long-standing bug**: sometimes a bug exists for a long time, and may have
known workarounds. [#3444][] is an example of a long-standing bug.

Where we feel that back-porting would be difficult or might be a stability
risk to clusters running cert-manager, we'll make the fix in a major
release but avoid back-porting the fix.

> #### Breaking changes {#breaking-changes}

Reviewer: nit: title case

```suggestion
#### Breaking Changes {#breaking-changes}
```

Author: Fixed in e05d5d2 (consisntent casing)

Breaking changes are changes that intentionally break the cert-manager
Kubernetes API or the command line flags. We avoid making breaking changes
where possible, and where they're required we'll give as much notice as
possible.

[#3393]: https://github.com/jetstack/cert-manager/issues/3393 "Broken CloudFlare DNS01 challenge"
[#2857]: https://github.com/jetstack/cert-manager/issues/2857 "CloudDNS DNS01 challenge crashes cert-manager"
[#3444]: https://github.com/jetstack/cert-manager/issues/3444 "Certificates do not get immediately updated after updating them"
[#3882]: https://github.com/jetstack/cert-manager/pull/3882: "Helm upgrade from v1.2 to v1.2 impossible due to a Helm bug"
[#3644]: https://github.com/jetstack/cert-manager/issues/3644 "Helm upgrade from v1.2 to v1.2 impossible due to a Helm bug"

> ## How we determine supported Kubernetes versions {#kubernetes-supported-versions}

Reviewer: nit: title case
```suggestion
## How We Determine Supported Kubernetes Versions {#kubernetes-supported-versions}
```

Author: Fixed in e05d5d2 (consisntent casing)

> The list of supported Kubernetes versions displayed in the section
> [Supported Releases](#supported-releases) depend on what the maintainers
> team thinks is reasonable to support and to test.

Reviewer: nit: grammar

```suggestion
The list of supported Kubernetes versions displayed in the
[Supported Releases](#supported-releases) section depends on
what the cert-manager maintainers think is reasonable to support
and to test.
```

Author: Fixed in de347d8, thanks!

[`master`]: https://testgrid.k8s.io/jetstack-cert-manager-master
[`next`]: https://testgrid.k8s.io/jetstack-cert-manager-next
[`previous`]: https://testgrid.k8s.io/jetstack-cert-manager-previous

## Terminology

The term "release" (or "minor release") refers to one minor version of
cert-manager. For example, 1.2 and 1.3 are two releases. Note that we do
not use the prefix `v` for releases (just "1.2"). This is because releases
are not used as git tags.

Patch releases use the `v` prefix (e.g., `v1.2.0`, `v1.3.1`...) since one
patch release = one git tag. The initial patch release is called "final
release":

| Type of release | Example of git tag | Corresponding release | Corresponding release branch\* |
| --------------- | ------------------ | --------------------- | ------------------------------ |
| Final release   | `v1.3.0`           | 1.3                   | `release-1.3`                  |
| Patch release   | `v1.3.1`           | 1.3                   | `release-1.3`                  |
| Pre-release     | `v1.4.0-alpha.0`   | N/A\*\*               | `release-1.4`                  |

\*For maintainers: each release has an associated long-lived branch that we

> call “release branch”. For example, `release-1.2` is the release branch for

Reviewer: nit: grammar
```suggestion
call the “release branch”. For example, `release-1.2` is the release branch for
```

Author: Fixed in de347d8, thanks!

release 1.2.

\*\*Pre-releases (e.g., `v1.3.0-alpha.0`) don't have a corresponding
release (e.g., 1.3) since a release only exists after a final release
(e.g., `v1.3.0`) has been created.

> Our naming scheme mostly follows the [Semantic Versioning

Author: Fixed in 17275b0 😅

> 2.0.0](https://semver.org/) with a `v` appended to the git tags and docker

Reviewer: nit: wrong word

```suggestion
2.0.0](https://semver.org/) with `v` prepended to git tags and docker
```

Author: Fixed in de347d8, thanks!

---

> title: "Supported releases"
> linkTitle: "Supported releases"

Reviewer: suggestion: title casing again; current titles of pages follow title casing

```suggestion
title: "Supported Releases"
linkTitle: "Supported Releases"
```

Author: Arghhh I just realized that the WHOLE website uses the Title Case convention for sections, which is (opinion!) a typographic mistake ([ref](https://uxplanet.org/why-letter-casing-is-important-to-consider-during-design-decisions-50402acd0a4e)) in the context of sections:

> Title Case is Inconsistent, Sometimes Even Incongruous. It Makes Sentences Little Difficult To Read As It Makes You Stop and Process Every Single Word. As If Giving Unnecessary Emphasis To Words. I Mean Why?

For the sake of consistency, I'll accept your suggestion 😇

weight: 5
type: "docs"
---

<!--
Inspired by https://istio.io/latest/about/supported-releases/
-->

This page lists the status, timeline and policy for currently supported

> releases.

Reviewer: praise: I cannot wait for this to be merged, it looks so useful for people. Thank you for writing it :+1:

| Final release   | `v1.3.0`           | 1.3                   | `release-1.3`                |
| Patch release   | `v1.3.1`           | 1.3                   | `release-1.3`                |
| Pre-release     | `v1.4.0-alpha.0`   | N/A\*                 | `release-1.4`                |

\*Pre-releases don't have a corresponding release since a release only
exists after a final release has been created.

Our naming scheme mostly follows the [Semantic Versioning
2.0.0](https://semver.org/) with a `v` appended to the git tags and docker
images:

```plain
v<major>.<minor>.<patch>
```

where `<minor>` is increased for each release, and `<patch>` counts the
number of patches for the current `<minor>` release. A patch is usually a
small change relative to the `<minor>` release.

## Support status of cert-manager releases

The period between two final releases is two months. The following table
presents the final releases for cert-manager.

| Version   | Support |   Release Date    |        EOL        | Supported Kubernetes versions | Untested but may work |
| --------- | ------- | :---------------: | :---------------: | :---------------------------: | :-------------------: |

> | master    | No      |                   |                   |                               |                       |

Author: Fixed in 90f00a15a4b221ee376cfa6c06bc4a3a69fdf818, you are right!

| [v1.4][]  | N/A     | [~][]Jun 11, 2021 |   [~][]Oct 2021   | 1.16, 1.17, 1.18, 1.19, 1.20  |                       |

> | [v1.3][]  | **Yes** |   Apr 08, 2021    |   [~][]Aug 2021   |    1.16, 1.17, 1.18, 1.19     |      1.20, 1.21       |

Author: Unfortunately, we are not testing with Kubernetes 1.20 as of today. This is configured in [cert-manager-release-previous-periodics.yaml](https://github.com/jetstack/testing/blob/a50d7471c0e39f584fcaca45e9d2eff75acab621/config/jobs/cert-manager/release-previous/cert-manager-release-previous-periodics.yaml).

We definitely need to add 1.20 and 1.21 to our "previous periodics"!

> | [v1.3][]  | **Yes** |   Apr 08, 2021    |   [~][]Aug 2021   |    1.16, 1.17, 1.18, 1.19     |      1.20, 1.21       |

Author: I removed the "Untested but may work" in 1827755  as discussed during the standup. I also added a "coverage table" with our current testing state. I also updated this table to reflect Irbe's jetstack/testing#487.

| [v1.2][]  | **Yes** |   Feb 10, 2021    | [~][]Jun 11, 2021 |          not tested           |      1.16 → 1.21      |
| [v1.1][]  | No      |   Nov 24, 2021    |   Apr 08, 2021    |          not tested           |      1.11 → 1.21      |
| [v1.0][]  | No      |   Sep 02, 2020    |   Feb 10, 2021    |          not tested           |      1.11 → 1.21      |
| [v0.16][] | No      |   Jul 23, 2020    |   Nov 24, 2020    |          not tested           |      1.11 → 1.21      |
| [v0.15][] | No      |   May 06, 2020    |   Sep 02, 2020    |          not tested           |      1.11 → 1.21      |
| [v0.14][] | No      |   Mar 11, 2020    |   Jul 23, 2020    |          not tested           |      1.11 → 1.21      |
| [v0.13][] | No      |   Jan 21, 2020    |   May 06, 2020    |          not tested           |      1.11 → 1.21      |
| [v0.12][] | No      |   Nov 27, 2019    |   Mar 11, 2020    |          not tested           |      1.11 → 1.21      |

> | [v0.11][] | No      |   Oct 10, 2019    |   Jan 21, 2020    |          not tested           |      1.9 → 1.21       |

Author: Yeah, totally agree, I don't want people to have to scroll down and skip "terminology" (which I am sure is useless, really, but it helped me a lot to be precise and define terms 😅) So, let's move "Terminology" at the bottom too.

Done in 90f00a1.

[v1.2]: https://cert-manager.io/docs/release-notes/release-notes-1.2
[v1.1]: https://cert-manager.io/docs/release-notes/release-notes-1.1
[v1.0]: https://cert-manager.io/docs/release-notes/release-notes-1.0
[v0.16]: https://cert-manager.io/docs/release-notes/release-notes-0.16
[v0.15]: https://cert-manager.io/docs/release-notes/release-notes-0.15
[v0.14]: https://cert-manager.io/docs/release-notes/release-notes-0.14
[v0.13]: https://cert-manager.io/docs/release-notes/release-notes-0.13
[v0.12]: https://cert-manager.io/docs/release-notes/release-notes-0.12
[v0.11]: https://cert-manager.io/docs/release-notes/release-notes-0.11

You can find available releases on the [releases
page](https://github.com/cert-manager/cert-manager/releases). You can find
the release notes for each minor release
[here](https://cert-manager.io/docs/release-notes/), and the upgrade
instructions are
[here](https://cert-manager.io/docs/installation/upgrading/). The
cert-manager release process is documented on the [release-process
page](https://cert-manager.io/docs/contributing/release-process/).

## Support policy

### What we mean by support

> Our support window is four month for each release branch (e.g.,

Author: Done in 479cfb9

`release-1.2`). That corresponds to the two latest releases, given that we

> produce a new final release every two months. We do offer two types of

Author: Done in 021d764, thanks

                     \                                             | SUPPORTED
                      \                                            | RELEASES
                       \         v1.2.0                            | = the two
                        \      Feb 10, 2021                        |   last
                         ------------+--------------> release-1.2  |   releases
                                      \                            v
                                       \
                                        \
                                         \
                                          -----------> master branch
                                                       April 1, 2021
```

### Technical support {#technical-support}

Technical assistance is offered on a best-effort basis on the two last
releases. Technical assistance is offered with the help of the community
and can be asked on the [Kubernetes Slack](https://slack.k8s.io/) (channel
`#cert-manager`), using [GitHub
Discussion](https://github.com/jetstack/cert-manager/discussions) or using
the [cert-manager-dev](https://groups.google.com/g/cert-manager-dev) Google

> group.

Author: Done in 7f90289, thank you

### Security and bug fixes {#bug-fixes-support}

Regarding the support of previous releases of cert-manager, we proceed to

> the back-porting of bug fixes depending on their criticality.

Author: Thanks!

Note that I'll use an em dash:

> We back-port important bug fixes — including security fixes — to all currently supported releases.

- [Security issues](#security-issues),
- [Critical bugs](#critical-bugs),
- [Long-standing bugs](#long-standing-bugs).

> The Kubernetes versions supported for each cert-manager release depend on

Author: Great suggestion, I moved it into its own section in da1b2a0.

the coverage of our nightly tests that we are able to maintain. Maintaining

> a large range of Kubernetes versions is tricky, that is why some versions

Author: Done in 28e0e04

are in the column "Untested but may work". The configuration for the
nightly can be seen in [`previous-periodics.yaml`][] and
[`next-periodics.yaml`][], and the results are visible at
[`testgrid-previous`][] and [`testgrid-next`][]).

[next-periodics.yaml]: https://github.com/jetstack/testing/blob/master/config/jobs/cert-manager/release-next/cert-manager-release-next-periodics.yaml
[previous-periodics.yaml]: https://github.com/jetstack/testing/blob/master/config/jobs/cert-manager/release-previous/cert-manager-release-previous-periodics.yaml
[testgrid-previous]: https://testgrid.k8s.io/jetstack-cert-manager-previous
[testgrid-next]: https://testgrid.k8s.io/jetstack-cert-manager-next

#### Security issues {#security-issues}

**Security issues** are fixed as soon as possible. They get back-ported to
the last two releases, and a new patch release is immediately created for
them:

> ```text

Author: Yeah, I could feel that the only really useful diagram is the "critical bugs" one. I removed it in 9e35628.

      \                                      commit  /
       \                                            /
        \                                          /
         \                                        /
          \         v1.2.0               v1.2.1  /
           ------------+---------------------+--/-----------> release-1.2
                        \           backport ^ /
                         \          commit  / /
                          \                / /
                           \              / /
                            \            / /
                             \          / /
                              ------------------------------> master
                                     ^   ^
                            security |   | two new patch
                               patch |   | releases
                              merged       created immediately
```

<!-- Diagram source: https://textik.com/#7c4096204b3c0ad3 -->

#### Critical bugs {#critical-bugs}

**Critical bugs** include both regression bugs as well as upgrade bugs.

Regressions are functionalities that did work in the previous release but

> no longer works. [#3393][] and [#2857][] are two examples of regressions.

Author: I added a paragraph:

> Note that intentional breaking changes do not belong to that category. Intentional breaking changes are changes that have been announced at least four months prior. For example, an intentional breaking change that is shipped in release 1.4 will first be announced in 1.2.

Not sure whether that belong to the section "Critical bugs" though 😅

> no longer works. [#3393][] and [#2857][] are two examples of regressions.

Author: Hmmmm I should first let the team decide if this is OK...

I will talk about this "breaking change policy" tomorrow morning during the standup

> Upgrade bugs are bugs (often Helm-related) preventing users from upgrading
> to earlier final releases of cert-manager. [#3882][] and [#3644][] are
> example of such upgrade bugs.

Reviewer: suggestion: grammar/rewording

```suggestion
Upgrade bugs are issues (often Helm-related) preventing users from upgrading
to currently supported releases from earlier releases of cert-manager. [#3882][] and [#3644][] are
examples of upgrade bugs.
```

> The fixes are back-ported; we then wait until the next final release for
> trigger a patch release. The patch release will be made for all supported
> release. Note that since we do only support the two last releases, patch
> releases are only made to the last release.

Reviewer: question: is this saying that we _don't_ create a new patch release after backporting the fix? if so, that seems wrong to me?

suggestion: grammar/rewording. I also personally find this quite confusing to read and parse, but I held off from suggesting wider changes to the wording since I don't fully understand the process we're describing here yet.

```suggestion
The fixes are back-ported and we then wait until the next final release before
triggering a patch release. The patch release will be made for all supported
releases, but note that since we only support the two last releases, patch
releases are only made to the last release.
```

Author: > is this saying that we don't create a new patch release after backporting the fix?

Yes, that's what @munnerz suggested in https://github.com/cert-manager/website/pull/452#discussion_r610496126; maybe I should just say this:

> The fixes related to critical bugs are back-ported "in batch": the patch release is created along with the next final release (we do not create a patch release for a single critical bug fix).

Author: I rephrased the paragraph in ae37d5e, let me know if you feel it is still confusing.

> In the below example, the release branches `release-1.2` and `release-1.3`

Reviewer: suggestion: wording

```suggestion
In the example below, the release branches `release-1.2` and `release-1.3`
```

Author: Fixed in ddb0e5e, thank you

are the two supported releases at the time of the release of 1.3.0, which

> means that the critical bug fix 1 and 2 will only be back-ported to

Reviewer: suggestion: wording/grammar

```suggestion
means that critical bug fixes 1 and 2 will only be back-ported to
```

Author: Fixed in 71647eee926968ebc20283a2bb108222ab9e7829

`release-1.2` and not to `release-1.1`:

```text
   v1.2.1                                v1.2.2
------+-------------------------------------+-----------> release-1.2
       \      backport ^  backport ^        ^
        \     commit  /   commit  /         | v1.2.2 is created
         \           /           /          | along with v1.3.0
          \         /           /
           \       /           /
            \     /           /          v1.3.0
             --------master-----------------+-----------> release-1.3
                 ^           ^               \
                 |           |                \
                 |           |                 \
           critical      critical               \
           bug fix 1     bug fix 2               \
           merged to     merged to                \
           master        master                    ------> master
```

<!-- Diagram source: https://textik.com/#7c4096204b3c0ad3 -->

> #### Long-standing bugs {#long-standing-bugs}

Reviewer: suggestion: I feel like this section could be really short and simple. The diagram here feels
less important than other diagrams too, since it basically just shows cert-manager's
normal development cycle.

An example of text that could work here:

```text
Sometimes a bug exists for a long time, and may have known workarounds. [#3444][] is an
example of a long-standing bug.

Where we feel that back-porting would be difficult or might be a stability risk to
clusters running cert-manager, we'll make the fix in a major release but avoid
back-porting the fix.
```

Author: Yeah, you are right, my phrasing is convoluted for no reason... I also removed the (useless) diagram too.

Done in c636091.

| Release branch | Prow configuration            | testgrid                | Kubernetes versions tested   | Periodicity |
| :------------: | :---------------------------- | :---------------------- | :--------------------------- | :---------: |
|     master     | [`next-periodics.yaml`][]     | [`testgrid-next`][]     | 1.16, 1.17, 1.18, 1.19, 1.20 | On each PR  |

> |  release-1.3   | [`previous-periodics.yaml`][] | [`testgrid-previous`][] | 1.16, 1.17, 1.18, 1.19, 1.20 | Every night |

Reviewer 4: > presubmits.yaml seems to be running Kubernetes 1.16 to 1.20 (not 1.21 though)... not sure though, I don't fully understand how the Prow config works sweat_smile

Good catch- you're right and it also made me realize that v1.16- v1.19 presubmits seem to not have run since April, see i.e https://testgrid.k8s.io/jetstack-cert-manager-presubmits-blocking#pull-cert-manager-e2e-v1-17. I am not entirely sure if we do want to run those. I can look into this.

> |  release-1.3   | [`previous-periodics.yaml`][] | [`testgrid-previous`][] | 1.16, 1.17, 1.18, 1.19, 1.20 | Every night |

Reviewer 4: Actually, we explicitly don't run v1.16- v1.19 on every PR, see https://github.com/jetstack/testing/blob/master/config/jobs/cert-manager/cert-manager-presubmits.yaml#L149

> |  release-1.3   | [`previous-periodics.yaml`][] | [`testgrid-previous`][] | 1.16, 1.17, 1.18, 1.19, 1.20 | Every night |

Author: Woah thanks for bringing that up, I had not noticed the \`always\_run: false\`!!

I'll fix the table 👍
